create_layout.py

```diff
@@ -88,51 +88,120 @@
 
-def create_material_rules(links, index):
-  """Create generic material rules (3 variants)
-
-  * MATCH available materials with products from previous step (links must be an
-  ordered list) and
-  * ALLOW available materials if it is the first step in the
-  list
-  Returns a list of material rules
-  NOTE: Read header docstring for ideas for more complexity.  """
-
-  expected_materials = []
-
-  if index == 0:
-    for material_name in links[index].materials.keys():
-      expected_materials.append(["ALLOW", material_name])
-    expected_materials.append(["DISALLOW", "*"])
-
-  else:
-    expected_materials = [
-        ["MATCH", "*", "WITH", "PRODUCTS", "FROM", links[index - 1].name]]
-
-  return expected_materials
-
-
-def create_product_rules(links, index):
-  """Create generic product rules (2 variants)
-
-  * ALLOW available products
-  * DISALLOW everything else
-
-  Returns a list of product rules
-  NOTE: Read header docstring for ideas for more complexity.  """
-
-
-  expected_products = []
-
-  for product_name in links[index].materials.keys():
-    expected_products.append(["ALLOW", product_name])
-
-  expected_products.append(["DISALLOW", "*"])
-
-  return expected_products
+def changes_between_snapshots(before_dict, after_dict):
+  """Given two 'snapshots' of an artifacts structure -- 'before' and 'after' --
+  return a tuple specifying which artifacts have been added, which have been
+  removed, which have been modified, and which have remained unchanged. Both
+  these dictionaries have artifact names as the keys and their hashes as the
+  values."""
+
+  before_set = set(before_dict.keys())
+  after_set = set(after_dict.keys())
+
+  removed_artifacts = before_set.difference(after_set)
+  added_artifacts = after_set.difference(before_set)
+
+  unchanged_artifacts = set()
+  modified_artifacts = set()
+  for key in before_set.intersection(after_set):
+    if before_dict[key] == after_dict[key]:
+      unchanged_artifacts.add(key)
+    else:
+      modified_artifacts.add(key)
+
+  return (unchanged_artifacts, modified_artifacts, added_artifacts,
+      removed_artifacts)
+
+def create_material_rules(previous_link, current_link):
+  """Create generic material rules
+
+  - MATCH available materials with products from previous step (links must be
+      an ordered list) and
+  - ALLOW available materials if it is the first step in the list
+  - DELETE removed materials
+
+  Args:
+    previous_link: a link of previous step, including previous step's materials
+        and products
+    current link: a link of current step, including current step's materials
+        and products
+
+  Returns:
+    a list of material rules
+  """
+
+  expected_materials_rules = []
+  unchanged_artifacts, modified_artifacts, _, deleted_artifacts = \
+      changes_between_snapshots(current_link.materials, current_link.products)
+  curr_step_materials = unchanged_artifacts.union(modified_artifacts,
+      deleted_artifacts)
```

Comment:
Is there a difference between `set(current_link.materials)` and `curr_step_materials`? If not, I'd prefer just using the former instead of generating the latter via union over `modified_artifacts` and `deleted_artifacts`.

Reply:
Yes, there is no difference.🤦 I have updated the changes.